source/workflow/app.py

```diff
@@ -538,4 +624,107 @@ def update_workflow_execution_status(id, status, message):
             InvocationType='Event'
         )
 
+# Find all of the execution error events for a state machine execution
+def get_execution_errors(arn):
+
+    try:
+
+        executions = []
+        paginator = SFN_CLIENT.get_paginator('get_execution_history')
+
+        # return all the history
+        page_iterator = paginator.paginate(executionArn=arn,
+                maxResults= 20,
+                reverseOrder=True)
+
+        # return only the history with exceptions
+        for page in page_iterator:
+            for event in page["events"]:
+              if (any(sub in event["type"] for sub in ['Failed', 'Aborted', 'TimedOut'])):
+                  logger.info("Found error event: {}".format(event))
+                  executions.append(event)
+
+        return executions
+
+    except:
+        logger.error("Unable to retrieve execution history for state machine termination")
+        raise
+
+# Find the earliest error message in the executions and retrieve additonal info if available
+def parse_execution_error(arn, executions, status):
+
+    message = "Caught Step Function Execution Status Change event for execution: "+ arn +", status:"+ status
+
+    # return the cause info from the most recent failure, if any
+    for execution in executions:
+      for key, value in execution.items():
+        if (any(sub in key for sub in ['Failed', 'Aborted', 'TimedOut'])):
+          if "cause" in value:
+              message = message+", cause: "+value["cause"]
+              break
+
+    return message
+
+# This lambda is invoked for Step Functions Execution Status Change events
+# that contain the status [ERROR, ABORTED, TIME_OUT].  Failures that occur
+# within the steps of a workflow state machine are handled by the
+# OperotorFailed lambda function, but if the state machine service throws
+# an exception that causes the state machine to terminate immediately,
+# it can't be handled within the state machine because the execution
+# is haulted.
+#
+# Info on events handled here:
+#     https://docs.aws.amazon.com/step-functions/latest/dg/cw-events.html
+#
+# Propagate the error and cause to the workflow control plane to close off
+# the workflow execution
+def workflow_error_handler_lambda (event, context):
+
+  try:
+
+    logger.info("workflow_error_handler_lambda: {}".format(json.dumps(event)))
+
+    if not ShortUUID:
+        raise Exception('ShortUUID is not set in lambda environment.')
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Using `Exception` and `BaseException` can make your code prone to errors and difficult to maintain. Instead, we recommend using one of the [Built-in Exceptions](https://docs.python.org/3/library/exceptions.html) or creating a custom exception class that is derived from `Exception` or one of its subclasses.



Similar issue at line numbers 693, 695, 697, and 699.

Reply:
We are not going to address this CodeGuru finding because we want to mirror the official AWS code sample for sigv4 signing: https://docs.aws.amazon.com/code-samples/latest/catalog/python-signv4-v4-signing-get-post.py.html.